### PR TITLE
Experiment with fully virtual VirtualThreadPool

### PR DESCRIPTION
**Pull request by gregw**
Virtual threads are used for all threading.

### CODE REVIEW

**Comment by gregw** [comment]
This will also need #11504

**Comment by sbordet** [comment]
If a selector thread is virtual, then it's pinned when it is waiting in `select()`.

It is difficult to show with `jstack`, but using a mix of `jstack` and a call to the MBean `HotspotDiagnostic.dumpThreads("/tmp/jvm.dump", JSON)`, and with system property `-Djdk.trackAllThreads=true`, you get, for example:

From `jstack`
```
"ForkJoinPool-1-worker-1" #31 [58094] daemon prio=5 os_prio=0 cpu=1.15ms elapsed=963.80s allocated=22128B defined_classes=5 tid=0x000074ee24d6f450  [0x000074edf482c000]
   Carrying virtual thread #30
	at jdk.internal.vm.Continuation.run(java.base@21.0.2/Continuation.java:248)
	at java.lang.VirtualThread.runContinuation(java.base@21.0.2/VirtualThread.java:221)
	at java.lang.VirtualThread$$Lambda/0x00000007c01ffa08.run(java.base@21.0.2/Unknown Source)
	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(java.base@21.0.2/ForkJoinTask.java:1423)
	at java.util.concurrent.ForkJoinTask.doExec(java.base@21.0.2/ForkJoinTask.java:387)
	at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(java.base@21.0.2/ForkJoinPool.java:1312)
	at java.util.concurrent.ForkJoinPool.scan(java.base@21.0.2/ForkJoinPool.java:1843)
	at java.util.concurrent.ForkJoinPool.runWorker(java.base@21.0.2/ForkJoinPool.java:1808)
	at java.util.concurrent.ForkJoinWorkerThread.run(java.base@21.0.2/ForkJoinWorkerThread.java:188)
```

From the `/tmp/jvm.dump` file:
```
         {
           "tid": "30",
           "name": "server0",
           "stack": [
              "java.base\/sun.nio.ch.EPoll.wait(Native Method)",
              "java.base\/sun.nio.ch.EPollSelectorImpl.doSelect(EPollSelectorImpl.java:121)",
              "java.base\/sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:130)",
              "java.base\/sun.nio.ch.SelectorImpl.select(SelectorImpl.java:147)",
              "org.eclipse.jetty.io@12.0.7-SNAPSHOT\/org.eclipse.jetty.io.ManagedSelector.nioSelect(ManagedSelector.java:181)",
              "org.eclipse.jetty.io@12.0.7-SNAPSHOT\/org.eclipse.jetty.io.ManagedSelector.select(ManagedSelector.java:188)",
              "org.eclipse.jetty.io@12.0.7-SNAPSHOT\/org.eclipse.jetty.io.ManagedSelector$SelectorProducer.select(ManagedSelector.java:612)",
              "org.eclipse.jetty.io@12.0.7-SNAPSHOT\/org.eclipse.jetty.io.ManagedSelector$SelectorProducer.produce(ManagedSelector.java:546)",
              "org.eclipse.jetty.util@12.0.7-SNAPSHOT\/org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.produceTask(AdaptiveExecutionStrategy.java:512)",
              "org.eclipse.jetty.util@12.0.7-SNAPSHOT\/org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:258)",
              "org.eclipse.jetty.util@12.0.7-SNAPSHOT\/org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.produce(AdaptiveExecutionStrategy.java:195)",
              "java.base\/java.util.concurrent.ThreadPerTaskExecutor$TaskRunner.run(ThreadPerTaskExecutor.java:314)",
              "java.base\/java.lang.VirtualThread.run(VirtualThread.java:309)"
           ]
         },
```

**Comment by sbordet** [comment]
`ForkJoinPool`, which is used as the default scheduler for virtual threads, can be constructed using a "parallelism" and a "maxPoolSize".

By default, parallelism=#cores and maxPoolSize=256 (see `VirtualThread.createDefaultScheduler()` in the JDK sources).

This means that if we have 8 cores and 8 selectors, we would be completely pinned, if not that FJP can "overshoot" and create other platform threads, up to 256 total.

Setting `-Djdk.virtualThreadScheduler.maxPoolSize=8` would limit the overshoot to 8, but if you have 8 virtual threads pinned in `select()` now the system is completely pinned.

The default idle timeout for the virtual thread FJP scheduler is 30 seconds, so they stay around for a while.

The option of only using virtual threads starts to feel in the camp of "magic", as in you think you know how the system works, but not really, until you know the details above.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/VirtualThreads.java [hunk 1]
```diff
@@ -125,15 +125,29 @@ public static boolean isVirtualThread()
      * @param namePrefix the prefix to use for the name of the virtual threads
      * @return a virtual threads {@code Executor} that will name the virtual threads according to the provided name prefix.
      */
-    public static Executor getNamedVirtualThreadsExecutor(String namePrefix)
+    public static ThreadFactoryExecutor getNamedVirtualThreadsExecutor(String namePrefix)
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/VirtualThreads.java]: This would be a breaking change (e.g. Spring).
Why is this necessary?

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java [hunk 2]
```diff
@@ -77,7 +77,7 @@
  * </ul>
  */
 @ManagedObject("A thread pool")
-public class QueuedThreadPool extends ContainerLifeCycle implements ThreadFactory, SizedThreadPool, Dumpable, TryExecutor, VirtualThreads.Configurable
+public class QueuedThreadPool extends ContainerLifeCycle implements ThreadPool, ThreadFactory, SizedThreadPool, Dumpable, TryExecutor, VirtualThreads.Configurable
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java]: Remove `ThreadPool` it already implements `SizedThreadPool`.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+{
+    private static final Logger LOG = LoggerFactory.getLogger(VirtualThreadPool.class);
+
+    private String _name = "vtp-%x".formatted(hashCode());
+    private VirtualThreads.ThreadFactoryExecutor _virtualExecutor;
+    private final AutoLock.WithCondition _joinLock = new AutoLock.WithCondition();
+
+    public VirtualThreadPool()
+    {
+        if (VirtualThreads.getDefaultVirtualThreadsExecutor() == null)
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java]: Use `VirtualThreads.areSupported()` instead.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+import org.eclipse.jetty.util.annotation.ManagedAttribute;
+import org.eclipse.jetty.util.component.ContainerLifeCycle;
+import org.eclipse.jetty.util.component.Dumpable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An implementation of {@link ThreadPool} interface that does not pool, but instead uses {@link VirtualThreads}.
+ */
+public class VirtualThreadPool extends ContainerLifeCycle implements ThreadFactory, ThreadPool, Dumpable, TryExecutor, VirtualThreads.Configurable
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java]: As virtual threads are not easily displayed in stack dumps, would be nice if this class could do it, under a configurable option, say "trackVirtualThreads" (which is similar to the JDK system property `-Djdk.trackAllThreads=true`).

With that option turned on, we can track the virtual thread when it executes the task, and remove it when the task is finished, so we can `dump()` the ones that we are tracking.

This is useful to dump the pinned virtual threads without having to know JDK system properties.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            throw new IllegalStateException(getState());
+        if (StringUtil.isBlank(name))
+            throw new IllegalArgumentException("Blank name");
+        _name = name;
+    }
+
+    @Override
+    protected void doStart() throws Exception
+    {
+        _virtualExecutor = VirtualThreads.getNamedVirtualThreadsExecutor(_name);
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java]: This is the "expensive" virtual thread `Executor` (has to concat strings to give the virtual thread a name), so there should be an option to use the default, rather than the named.
Perhaps just `if (_name == null || _name.isBlank())`.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        while (isStopping())
+        {
+            Thread.onSpinWait();
+        }
+    }
+
+    @Override
+    public int getThreads()
+    {
+        return -1;
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java]: See comment above, we can track virtual threads.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/VirtualThreads.java (hunk 1 above)
Comment (joakime) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/VirtualThreads.java]: I cannot find any code at spring-projects that use the method `getNamedVirtualThreadsExecutor`

https://github.com/search?q=org%3Aspring-projects+jetty+getNamedVirtualThreadsExecutor&type=code

This includes spring-boot, spring-framework, examples, etc ...

**Comment by lorban** [comment]
We should test this change with the perf profiler but #11513 should be fixed first.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/VirtualThreads.java (hunk 1 above)
Comment (sbordet) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/VirtualThreads.java]: @joakime see discussion in #11353.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java (hunk 4 above)
Comment (gregw) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java]: I'll do this if we decide this is actually something we want to merge.

**Comment by gregw** [comment]
@sbordet @lorban because of pinning, I doubt this is a viable mode to run it.... but it would still be interesting to include this branch benchmarks just to see what a fully virtual server would run like.   This would remove the overheads of both QTP and RTE, but with the risk of parallel slowdown.   It might still be a fast (if somewhat risky) mode to run in?
If we ever merge this, then we probably should modify the selector heuristics to never have more selectors than cores.

**Comment by gregw** [comment]
Considering pinning, I think the current virtual thread mode is safe because we force reserved threads to 0, thus we will never tryExecute a producer task, so the producer should never be virtual.  We only use virtual threads to consume BLOCKING tasks, which should not be selectors.... but may be other code that pins the cores, but that should not be us at least.

**Comment by gregw** [comment]
@sbordet @lorban I reverted the name `threadpool-hybrid` to `threadpool-virtual`.  The new mode is now called `threadpool-all-virtual`.

**Comment by gregw** [comment]
@sbordet @lorban thoughts on this?   Is this something we want to play with, or is pinning just too much of an issue to make it worthwhile?

**Comment by sbordet** [comment]
@gregw keep it! 😄 

Various comments:

1. We tried to debug a virtual thread issue using what recommended by by [JEP 444](https://openjdk.org/jeps/444) in section `Observing virtual threads`.
Unfortunately, `jcmd` outputs a limited dump that does not report enough information (e.g. does not associate the virtual thread with its carrier, so there is no information if a virtual thread is pinned).
The `TrackingVirtualExecutor` in this PR will be able to do so, provided that dumping capabilities are added. 
I have just added this:
```java
@Override
public void dump(Appendable out, String indent) throws IOException
{
    dumpObjects(out, indent, new DumpableCollection("virtual threads", _threads.stream().map(t -> new DumpableCollection(t.toString(), List.of(t.getStackTrace()))).toList()));
}
``` 
to get state & carrier information, and stack frames even of parked virtual threads.
This would be invaluable.
However, `TrackingVirtualExecutor` should be a reusable class that I can use when doing:
```java
QueuedThreadPool qtp = new QueuedThreadPool();
qtp.setVirtualThreadExecutor(new TrackingVirtualExecutor(Executors.newThreadPerTaskExecutor()));
```
2. I'd get rid of `ThreadFactoryExecutor` to use the established patter that if you need a `ThreadFactory`, it is passed as a parameter.
3. By default, carrier threads can overshoot up to 256, unless explicitly specified with system property `jdk.virtualThreadScheduler.maxPoolSize`.
Therefore, I think there may be value to have a `VirtualThreadPool` (although the name is an anti-pattern 😄) as, by default, even if we pin virtual threads that do `select()`, the JDK overshoots up to 256 so there should be room to spare for carriers, even when setting `selectors=cores`, which would pin all "default" virtual threads, but the JDK would still allocate more carriers (unless you have 256 cores).
4. I'd rename the `-all-` with `-full-`.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/VirtualThreads.java [hunk 7]
```diff
@@ -233,4 +257,7 @@ default void setUseVirtualThreads(boolean useVirtualThreads)
     private VirtualThreads()
     {
     }
+
+    public interface ThreadFactoryExecutor extends ThreadFactory, Executor
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/VirtualThreads.java]: Not a big fan of this... if it can be kept separated, I think it's better.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+        if (StringUtil.isBlank(name) && name != null)
+            throw new IllegalArgumentException("Blank name");
+        _name = name;
+    }
+
+    public void setTracking(boolean tracking)
+    {
+        if (isRunning())
+            throw new IllegalStateException(getState());
+        _threads = tracking ? Collections.newSetFromMap(new WeakHashMap<>()) : null;
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java]: Not thread safe, and no need to be weak.
Just use `ConcurrentHashMap.newKeySet()` or similar.

Also, what if `tracking=false`?

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+        return false;
+    }
+
+    @Override
+    public void execute(Runnable task)
+    {
+        _virtualExecutor.execute(task);
+    }
+
+    private class TrackingVirtualExecutor implements VirtualThreads.ThreadFactoryExecutor
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java]: Should be factored out for reusability, move `_threads` into this class, and add dump capabilities.

**Review by sbordet — CHANGES_REQUESTED** [review]
Just the virtual thread tracking capabilities are a must-have.
The rest needs cleanup IMHO, but let's continue on this PR.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java (hunk 8 above)
Comment (gregw) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java]: then _threads is set to null

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java (hunk 9 above)
Comment (gregw) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java:204]: I think we should leave it here until there is a need to reuse.
I have moved _threads and done other cleanups.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/VirtualThreads.java (hunk 7 above)
Comment (gregw) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/VirtualThreads.java]: removed, but we are no longer able to implement a Virtual ThreadFactory

**Comment by gregw** [comment]
@olamy the CI build is [green](https://jenkins.webtide.net/blue/organizations/jenkins/jetty.project/detail/PR-11501/19/pipeline).  Any idea how I can make it update here?

File: jetty-core/jetty-server/src/main/config/etc/jetty-threadpool-all-virtual.xml [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+<Configure>
+  <New id="threadPool" class="org.eclipse.jetty.util.thread.VirtualThreadPool">
+    <Set name="name" property="jetty.threadPool.namePrefix" />
+  </New>
+
+  <Call class="org.slf4j.LoggerFactory" name="getLogger">
+    <Arg>org.eclipse.jetty</Arg>
+    <Call name="info">
+      <Arg>All Virtual threads are enabled.</Arg>
```
Comment (sbordet) [inline jetty-core/jetty-server/src/main/config/etc/jetty-threadpool-all-virtual.xml]: Should be more clear as in "Virtual threads enabled, using only virtual threads".

Also the file name should replace "-all-" with "-full-" or "-only-".

File: jetty-core/jetty-server/src/main/config/modules/threadpool-all-virtual.mod [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+[xml]
+etc/jetty-threadpool-all-virtual.xml
+
+[ini-template]
+# tag::documentation[]
+## Platform threads name prefix.
+#jetty.threadPool.namePrefix=vtp<hashCode>
+
+# end::documentation[]
```
Comment (sbordet) [inline jetty-core/jetty-server/src/main/config/modules/threadpool-all-virtual.mod:19]: Ditto file name change.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/TrackingExecutor.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executor;
+
+import org.eclipse.jetty.util.annotation.ManagedAttribute;
+import org.eclipse.jetty.util.annotation.ManagedObject;
+import org.eclipse.jetty.util.component.Dumpable;
+
+@ManagedObject("Tracking Executor wrapper")
+class TrackingExecutor implements Executor, Dumpable
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/TrackingExecutor.java]: Should be `public`.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/TrackingExecutor.java [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    @Override
+    public void execute(Runnable task)
+    {
+        _threadFactoryExecutor.execute(() ->
+        {
+            try
+            {
+                _threads.add(Thread.currentThread());
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/TrackingExecutor.java]: Introduce variable for `Thread.currentThread()` so it is reused in the `finally` block.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+    }
+
+    public void setTracking(boolean tracking)
+    {
+        if (isRunning())
+            throw new IllegalStateException(getState());
+        _tracking = tracking;
+    }
+
+    public boolean isTracking()
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java]: Javadocs and `@ManagedAttribute`.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+                        l.await();
+                    }
+                }
+                catch (InterruptedException e)
+                {
+                    throw new RuntimeException(e);
+                }
+            }
+        };
+        _main.start();
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java:126]: Is this platform thread only necessary to keep the server alive?
The thread name should not contain "virtual" because it's not virtual. How about "keepalive"?

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+        addBean(_virtualExecutor);
+        super.doStart();
+    }
+
+    @Override
+    protected void doStop() throws Exception
+    {
+        super.doStop();
+        _virtualExecutor = null;
+        _main = null;
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java:148]: I think we must call `removeBean(_virtualExecutor)` to pair the add on `doStart()`.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+    @Override
+    public Executor getVirtualThreadsExecutor()
+    {
+        return _virtualExecutor;
+    }
+
+    @Override
+    public void setVirtualThreadsExecutor(Executor executor)
+    {
+        throw new UnsupportedOperationException("cannot set VirtualThreadExecutor");
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java]: Why not?
Users may pass their own if so wish (e.g. shared with other components not related to Jetty), and we can wrap it.
Perhaps just have a guard for `isStarted()`?

File: jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/thread/VirtualThreadPoolTest.java [hunk 18]
```diff
@@ -0,0 +1,202 @@
...
+                    throw new IllegalStateException();
+
+            }
+            catch (InterruptedException e)
+            {
+                throw new RuntimeException(e);
+            }
+            finally
+            {
+                System.err.println("RAN!" + Thread.currentThread());
```
Comment (sbordet) [inline jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/thread/VirtualThreadPoolTest.java]: Leftover?

**Review by sbordet — CHANGES_REQUESTED** [review]

**Comment by olamy** [comment]
> @olamy the CI build is [green](https://jenkins.webtide.net/blue/organizations/jenkins/jetty.project/detail/PR-11501/19/pipeline). Any idea how I can make it update here?

fixed.

File: jetty-core/jetty-server/src/main/config/etc/jetty-threadpool-all-virtual.xml (hunk 10 above)
Comment (gregw) [reply jetty-core/jetty-server/src/main/config/etc/jetty-threadpool-all-virtual.xml]: I prefer "all".   "full" is confusing when pools are involved. What does it mean to have a full non pooling pool?  "Only" is ok, but it is just as easy to say "the threads used are all virtual", as it is to say "the threads used are only virtual". However, "only virtual" can be taken as a pejorative  (which in this case might be appropriate :), so I prefer "all"

File: jetty-core/jetty-server/src/main/config/modules/threadpool-all-virtual.mod (hunk 11 above)
Comment (gregw) [reply jetty-core/jetty-server/src/main/config/modules/threadpool-all-virtual.mod:19]: Ditto I prefer "all"

**Comment by gregw** [comment]
@lorban @sbordet nudge!

File: jetty-core/jetty-server/src/main/config/etc/jetty-threadpool-all-virtual.xml [hunk 19]
```diff
@@ -0,0 +1,15 @@
...
+
+<Configure>
+  <New id="threadPool" class="org.eclipse.jetty.util.thread.VirtualThreadPool">
+    <Set name="name" property="jetty.threadPool.namePrefix" />
+  </New>
+
+  <Call class="org.slf4j.LoggerFactory" name="getLogger">
+    <Arg>org.eclipse.jetty</Arg>
+    <Call name="info">
+      <Arg>Virtual threads enabled. Using all virtual threads.</Arg>
```
Comment (sbordet) [inline jetty-core/jetty-server/src/main/config/etc/jetty-threadpool-all-virtual.xml]: Still not clear to me... seems there is a bunch of threads, and I'm using them all.

How about "Virtual threads available, using virtual threads for all tasks."

In the other we can say "Virtual threads available, using virtual threads only for application tasks."

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/TrackingExecutor.java [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+import org.eclipse.jetty.util.component.Dumpable;
+
+@ManagedObject("Tracking Executor wrapper")
+public class TrackingExecutor implements Executor, Dumpable
+{
+    private final Executor _threadFactoryExecutor;
+    private final Set<Thread> _threads = ConcurrentHashMap.newKeySet();
+    private boolean _detailed;
+
+    TrackingExecutor(Executor executor, boolean detailed)
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/TrackingExecutor.java]: Must be public.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java (hunk 14 above)
Comment (sbordet) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java]: Also, move it above the setter, since the other have the getter method before the setter method.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+    {
+        _detailedDump = detailedDump;
+        if (_virtualExecutor instanceof TrackingExecutor trackingExecutor)
+            trackingExecutor.setDetailedDump(detailedDump);
+    }
+    
+    @Override
+    protected void doStart() throws Exception
+    {
+        _main = new Thread("keepalive")
```
Comment (lorban) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java]: This thread deserves a better name. How about `jetty-virtual-thread-pool-keepalive`?

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java [hunk 22]
```diff
@@ -0,0 +1,220 @@
...
+    {
+        super.doStop();
+        removeBean(_virtualExecutor);
+        if (!_externalExecutor)
+            _virtualExecutor = null;
+        _main = null;
+
+        try (AutoLock.WithCondition l = _joinLock.lock())
+        {
+            l.signalAll();
```
Comment (lorban) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java:152]: `l.signal()` seems enough here.

**Review by lorban — DISMISSED** [review]

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java (hunk 22 above)
Comment (gregw) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java:152]: ya'd think! But the join test fails with just signal, I assume because multiple threads can join and wait.

**Comment by gregw** [comment]
@sbordet sorry I missed a few comments.  I'm having troubling making the new intellij review mode update always.  I think I have got them all now.

File: jetty-core/jetty-server/src/main/config/etc/jetty-threadpool-all-virtual.xml [hunk 23]
```diff
@@ -0,0 +1,15 @@
...
+
+<Configure>
+  <New id="threadPool" class="org.eclipse.jetty.util.thread.VirtualThreadPool">
+    <Set name="name" property="jetty.threadPool.namePrefix" />
+  </New>
+
+  <Call class="org.slf4j.LoggerFactory" name="getLogger">
+    <Arg>org.eclipse.jetty</Arg>
+    <Call name="info">
+      <Arg>Virtual threads enabled. Using virtual threads for all Jetty tasks.</Arg>
```
Comment (sbordet) [inline jetty-core/jetty-server/src/main/config/etc/jetty-threadpool-all-virtual.xml]: I'd remove "Jetty" here; it is redundant and feels like "only the tasks of the Jetty implementation, so not those of the application".

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java [hunk 24]
```diff
@@ -0,0 +1,226 @@
...
+    {
+        if (isRunning())
+            throw new IllegalStateException(getState());
+        if (StringUtil.isBlank(name) && name != null)
+            throw new IllegalArgumentException("Blank name");
+        _name = name;
+    }
+
+    /**
+     *
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/VirtualThreadPool.java]: Empty line.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-server/src/main/config/etc/jetty-threadpool-all-virtual.xml (hunk 23 above)
Comment (gregw) [reply jetty-core/jetty-server/src/main/config/etc/jetty-threadpool-all-virtual.xml]: @sbordet I don't agree.  We frequently have issues with users that use a mix of scheduling mechanisms.   They might have an asynchronous Jetty Handler that uses some 3rd party async library that ends up using platform threads for its callback, then end up calling Jetty callbacks.    It is important to be clear that we are not making all threads in the JVM virtual, just the ones that we dispatch.

Comment (sbordet) [reply jetty-core/jetty-server/src/main/config/etc/jetty-threadpool-all-virtual.xml]: @gregw this is the `VirtualThreadPool` and "all tasks" refers to the ones submitted to this pool.
There is no need to say "Jetty", it's just the task submitted to this pool.

Comment (sbordet) [reply jetty-core/jetty-server/src/main/config/etc/jetty-threadpool-all-virtual.xml]: A server uses a platform thread pool, but have a proxy web application using `HttpClient` that uses `VirtualThreadPool` and now we have "all Jetty tasks" but which ones, the client's or the server's?

Perhaps let's add "Virtual thread available. Using virtual threads for all tasks submitted to VirtualThreadPool@1234."

**Review by sbordet — CHANGES_REQUESTED** [review]
See comment.

File: jetty-core/jetty-server/src/main/config/etc/jetty-threadpool-all-virtual.xml (hunk 10 above)
Comment (joakime) [reply jetty-core/jetty-server/src/main/config/etc/jetty-threadpool-all-virtual.xml]: I like `jetty-threadpool-only-virtual.xml`

**Review by joakime — COMMENTED** [review]
Generally I'm good with this.
Just some naming niggles, but @sbordet is on the same page as I am there.
Perhaps this should be done for Jetty 12.1.0?  That would encourage people to update to 12.1.0 as well.

File: jetty-core/jetty-server/src/main/config/etc/jetty-threadpool-all-virtual.xml (hunk 23 above)
Comment (gregw) [reply jetty-core/jetty-server/src/main/config/etc/jetty-threadpool-all-virtual.xml]: I still don't like that, as it is unclear what tasks are submitted to the VTP.   It is kind of saying the obvious that the VTP will use virtual threads.   We need to make the distinction between using threads for tasks calling the application vs tasks for all jetty tasks.    I think the current text does that OK.

Comment (sbordet) [reply jetty-core/jetty-server/src/main/config/etc/jetty-threadpool-all-virtual.xml]: @gregw I don't like it because there is no definition of what a "Jetty task" is, and I don't think we want to define that as it won't be useful to users.

We just need to make a difference between: everything is run by a virtual thread, or only web application code, so specifying "Jetty" is redundant because _all_ tasks will be run by virtual threads, not only the "Jetty" tasks.

Comment (gregw) [reply jetty-core/jetty-server/src/main/config/etc/jetty-threadpool-all-virtual.xml]: OK, but you will get a great big "I told you so" the first time we get somebody confused that they have enabled this module, but that there are platform threads calling jetty callbacks and doing jetty things.

**Review by sbordet — APPROVED** [review]
